pkg/service/scheduler/policy.go

```diff
@@ -13,55 +14,85 @@ import (
 
 // Policy decides if given task can be run.
 type Policy interface {
-	PreRun(clusterID, taskID, runID uuid.UUID) error
-	PostRun(clusterID, taskID, runID uuid.UUID)
+	PreRun(clusterID, taskID, runID uuid.UUID, taskType TaskType) error
+	PostRun(clusterID, taskID, runID uuid.UUID, taskType TaskType)
 }
 
 // PolicyRunner is a runner that uses policy to check if a task can be run.
 type PolicyRunner struct {
 	Policy Policy
 	Runner Runner
+
+	// TaskType of a task that this runner is executing
+	TaskType TaskType
 }
 
 // Run implements Runner.
 func (pr PolicyRunner) Run(ctx context.Context, clusterID, taskID, runID uuid.UUID, properties json.RawMessage) error {
-	if err := pr.Policy.PreRun(clusterID, taskID, runID); err != nil {
+	if err := pr.Policy.PreRun(clusterID, taskID, runID, pr.TaskType); err != nil {
 		return err
 	}
-	defer pr.Policy.PostRun(clusterID, taskID, runID)
+	defer pr.Policy.PostRun(clusterID, taskID, runID, pr.TaskType)
 	return pr.Runner.Run(ctx, clusterID, taskID, runID, properties)
 }
 
 var errClusterBusy = errors.New("another task is running")
 
-// LockClusterPolicy is a policy that can execute only one task at a time.
-type LockClusterPolicy struct {
-	mu   sync.Mutex
-	busy map[uuid.UUID]struct{}
+// TaskExclusiveLockPolicy is a policy that executes the exclusiveTask only if there are no other tasks in the cluster.
+// Conversely, other tasks can run only if the exclusiveTask is not running.
+// Additionally this policy ensures that only one task of a task type can be executed at a time in a cluster.
+type TaskExclusiveLockPolicy struct {
+	mu      sync.Mutex
+	running map[uuid.UUID]map[TaskType]struct{}
+
+	exclusiveTask TaskType
 }
 
-func NewLockClusterPolicy() *LockClusterPolicy {
-	return &LockClusterPolicy{
-		busy: make(map[uuid.UUID]struct{}),
+func NewTaskExclusiveLockPolicy(exclusiveTask TaskType) *TaskExclusiveLockPolicy {
+	return &TaskExclusiveLockPolicy{
+		running: map[uuid.UUID]map[TaskType]struct{}{},
+
+		exclusiveTask: exclusiveTask,
 	}
 }
 
-// PreRun implements Policy.
-func (p *LockClusterPolicy) PreRun(clusterID, _, _ uuid.UUID) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+// PreRun acquires exclusive lock on a cluster for a provided taskType.
+func (t *TaskExclusiveLockPolicy) PreRun(clusterID, _, _ uuid.UUID, taskType TaskType) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	cluster, ok := t.running[clusterID]
+	if !ok {
+		t.running[clusterID] = map[TaskType]struct{}{}
+	}
+	if len(cluster) == 0 {
+		t.running[clusterID][taskType] = struct{}{}
+		return nil
+	}
 
-	if _, ok := p.busy[clusterID]; ok {
+	// Exclusive task can be run only when no other tasks is running.
+	if taskType == t.exclusiveTask {
+		return fmt.Errorf("run exclusive task %s: %w", taskType, errClusterBusy)
+	}
+
+	// Any other task can't be run when exclusive task is running.
+	if _, ok := cluster[t.exclusiveTask]; ok {
```

Comment:
I agree that this code is a bit confusing, so I'll refactor it :) 
jfyi: cluster couldn't be nil here, because after `!ok` we have a condition `len(cluster) == 0` that will catch`nil` cluster.  But yeah, it's confusing